zkp/js/test/check_nullifier_value.js

```diff
@@ -0,0 +1,221 @@
+    const output1 = poseidonHash([BigInt(outputValues[0]), salt3, ...Alice.pubKey]);
+    const outputCommitments = [output1];
+
+    const witness = await circuit.calculateWitness(
+      {
+        nullifiers,
+        inputCommitments,
+        inputValues,
+        inputSalts: [salt1, 0],
+        inputOwnerPrivateKey: senderPrivateKey,
+        root: proof1.root.bigInt(),
+        merkleProof: [proof1.siblings.map((s) => s.bigInt()), proof2.siblings.map((s) => s.bigInt())],
+        enabled: [1, 0],
+        outputCommitments,
+        outputValues,
+        outputSalts: [salt3],
+        outputOwnerPublicKeys: [Alice.pubKey],
+      },
+      true
+    );
+
+    expect(witness[1]).to.equal(BigInt(62));
+    expect(witness[2]).to.equal(BigInt(nullifiers[0]));
+    expect(witness[3]).to.equal(BigInt(nullifiers[1]));
+    expect(witness[4]).to.equal(proof1.root.bigInt());
+  });
+
+  it('should generate a valid proof that can be verified successfully', async () => {
+    const inputValues = [15, 100];
+    const outputValues = [35];
+
+    // create two input UTXOs, each has their own salt, but same owner
+    const senderPrivateKey = formatPrivKeyForBabyJub(Alice.privKey);
+    const salt1 = newSalt();
+    const input1 = poseidonHash([BigInt(inputValues[0]), salt1, ...Alice.pubKey]);
+    const salt2 = newSalt();
+    const input2 = poseidonHash([BigInt(inputValues[1]), salt2, ...Alice.pubKey]);
+    const inputCommitments = [input1, input2];
+
+    // create the nullifiers for the input UTXOs
+    const nullifier1 = poseidonHash3([BigInt(inputValues[0]), salt1, senderPrivateKey]);
+    const nullifier2 = poseidonHash3([BigInt(inputValues[1]), salt2, senderPrivateKey]);
+    const nullifiers = [nullifier1, nullifier2];
+
+    // calculate the root of the SMT
+    await smtAlice.add(input1, input1);
+    await smtAlice.add(input2, input2);
+
+    // generate the merkle proof for the inputs
+    const proof1 = await smtAlice.generateCircomVerifierProof(input1, ZERO_HASH);
+    const proof2 = await smtAlice.generateCircomVerifierProof(input2, ZERO_HASH);
+
+    // create two output UTXOs, they share the same salt, and different owner
+    const salt3 = newSalt();
+    const output1 = poseidonHash([BigInt(outputValues[0]), salt3, ...Alice.pubKey]);
+    const outputCommitments = [output1];
+
+    const startTime = Date.now();
+    const witness = await circuit.calculateWTNSBin(
+      {
+        nullifiers,
+        inputCommitments,
+        inputValues,
+        inputSalts: [salt1, salt2],
+        inputOwnerPrivateKey: senderPrivateKey,
+        root: proof1.root.bigInt(),
+        merkleProof: [proof1.siblings.map((s) => s.bigInt()), proof2.siblings.map((s) => s.bigInt())],
+        enabled: [1, 1],
+        outputCommitments,
+        outputValues,
+        outputSalts: [salt3],
+        outputOwnerPublicKeys: [Alice.pubKey],
+      },
+      true
+    );
+
+    const { proof, publicSignals } = await groth16.prove(provingKeyFile, witness);
```

Comment:
the output UTXOs would be added to the local merkle tree after the transaction has been confirmed, because only at that point would the output UTXOs be added to the merkle tree maintained by the smart contract.

in the circuit tests, we stop short of submitting transactions. once we are able to successfully generate a proof, and are able to verify it locally (in node.js or golang snark SDKs), it proves that the circuit has been designed correctly.